### PR TITLE
Support for the instructions of ZX Spectrum Next

### PR DESCRIPTION
**Pull request by mborik**
I found this active fork without pull request, but it's interesting and related to #19 feature request.
Tasks to PR acceptance:
- [ ] it should be tested
- [ ] cleanup of typos and strange formatting

### CODE REVIEW

File: sjasm/z80.cpp [hunk 1]
```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
...
+        EmitByte(0xED);
+        EmitByte(0x93);
+    }
+
+    void OpCode_PIXELAD() {
+        EmitByte(0xED);
+        EmitByte(0x94);
+    }
+
+    void OpCode_FILLDE() {
```
Comment (ped7g) [inline sjasm/z80.cpp]: `fillde` was removed from Z80N specs (sad but true)

File: sjasm/z80.cpp [hunk 2]
```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
...
+        EmitByte(0xED);
+        EmitByte(0xAC);
+    }
+
+    void OpCode_LDDRX() {
+        EmitByte(0xED);
+        EmitByte(0xBC);
+    }
+
+    void OpCode_LDIRSCALE() {
```
Comment (ped7g) [inline sjasm/z80.cpp]: Not implemented in Z80N yet, and seems now very unlikely to happen

File: sjasm/z80.cpp [hunk 3]
```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
...
+    }
+
+
+
+    void OpCode_OUTINB() {
+        EmitByte(0xED);
+        EmitByte(0x90);
+    }
+
+    void OpCode_POPX() {
```
Comment (ped7g) [inline sjasm/z80.cpp]: :dagger:  RIP :) (`inc sp` `inc sp` is not only same length, but faster :D )

File: sjasm/z80.cpp [hunk 4]
```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
...
+        EmitByte(0xED);
+        EmitByte(0x8B);
+    }
+
+    void OpCode_SETAE() {
+        EmitByte(0xED);
+        EmitByte(0x95);
+    }
+
```
Comment (ped7g) [inline sjasm/z80.cpp]: And this extension was done based on some older core 1.xx versions, as it's missing these instructions:
`push imm16` (BTW current `OpCode_PUSH` implementation can lead to buffer overflow - not part of this changeset, but needs fix)
`bsla | bsra | bsrl | bsrf | brlc de,b` and `jp (c)`

**Comment by ped7g** [comment]
BTW as an sjasmplus outsider, the #19 was read by me as "we don't even want it"... if it was more like "Ain't Nobody Got Time For That", I would definitely love to see Z80N being supported by z00m128 upstream repository.

File: sjasm/z80.cpp (hunk 4 above)
Comment (ped7g) [reply sjasm/z80.cpp]: I take back the buffer overflow comment about current OpCode_PUSH, it's actually protected... but I'm probably going to modify it in the similar way how other opcodes deal with multi-operands, as part of the Next fixes...

File: sjasm/z80.cpp [hunk 5]
```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
...
+            }
+            else{
+                e[0] = 0xed; e[1] = 0x91;e[2] = GetByte(lp);
+                if(!comma(lp)){
+                    break;
+                }
+                e[3] = GetByte(lp);
+            }
+            EmitBytes(e);
+            /* (begin add) */
```
[the remark this answers is not in the document]
Comment (ped7g) [reply sjasm/z80.cpp]: resolved: they are really just pre-VCS era type of comments and can be safely ignored/removed

**Comment by CatpainBlack** [comment]
Just a quick note, the code in my branch is a quick hack, I'm planning on revisiting it and fixing up the Z80N instructions in the not too distant future.

**Comment by ped7g** [comment]
> Just a quick note, the code in my branch is a quick hack, I'm planning on revisiting it and fixing up the Z80N instructions in the not too distant future.

Well, all fixed in my branch now, review + testing is welcome and needed.
Also regarding sjasmplus vs Next, there's more to it than just instructions, but I don't think I will work on the DEVICE and other directives in near future, as it's of low priority for my current projects.

**Comment by CatpainBlack** [comment]
> > Just a quick note, the code in my branch is a quick hack, I'm planning on revisiting it and fixing up the Z80N instructions in the not too distant future.
> 
> Well, all fixed in my branch now, review + testing is welcome and needed.
> Also regarding sjasmplus vs Next, there's more to it than just instructions, but I don't think I will work on the DEVICE and other directives in near future, as it's of low priority for my current projects.

I'm happy to help out, I'm using sjasmplus as main main assembler for Next development

**Comment by mborik** [comment]
Closing this pull-request, because of Ped's PR #29 where all requested changes has been applied, followed by massive refactor.

**Pull request closed without merging**